cylc/flow/xtrigger_mgr.py

```diff
@@ -357,16 +356,28 @@ def callback(self, ctx: SubFuncContext):
         if satisfied:
             self.data_store_mgr.delta_task_xtrigger(sig, True)
             LOG.info('xtrigger satisfied: %s = %s', ctx.label, sig)
-            self.pflag = True
             self.sat_xtrig[sig] = results
 
-    def check_xtriggers(self, itasks: List[TaskProxy]):
-        """See if any xtriggers are satisfied.
+    def check_xtriggers(
+            self,
+            itasks: List[TaskProxy],
+            db_update_func: Callable[[dict], None]) -> Set[TaskProxy]:
+        """Check if any of itasks' xtriggers have become satisfied.
+
+        Return set of newly satisfied tasks.
 
         Args:
-            itasks (List[TaskProxy]): list of TaskProxy's
+            itasks: list of tasks to check
+            db_update_func: method to update xtriggers in the DB
+
         """
-        self.collate(itasks)
-        for itask in itasks:
-            if itask.state.xtriggers:
-                self.satisfy_xtriggers(itask)
+        satisfied = set(
+            [
+                itask for itask in itasks
+                if itask.state.xtriggers_all_satisfied()
+            ]
+        )
```

Comment:
```suggestion
        satisfied = set(
            itask for itask in itasks
            if itask.state.xtriggers_all_satisfied()
        )
```

Reply:
Or better still use a native set comprehension `satisfied = { itask for itask ...`

Reply:
Doh, welcome to Python 3 :rofl: